src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala

```diff
@@ -182,15 +191,30 @@ class IpLookups(
   private def performLookupsWithLruCache(
     lru: SynchronizedLruMap[String, IpLookupResult],
     ip: String
-  ): IpLookupResult = lru.get(ip) match {
-    case Some(result) => result // In the LRU cache
-    case None => // Not in the LRU cache
-      val result = performLookupsWithoutLruCache(ip)
-      lru.put(ip, result)
-      result
+  ): IO[IpLookupResult] = {
+    val lookupAndCache = for {
+      result <- performLookupsWithoutLruCache(ip)
+      _      <- putFromMap(lru, ip, result)
+    } yield result
+
+    getFromMap(lru, ip)
+      .map(_.map(IO.pure(_)))
+      .flatMap(_.getOrElse(lookupAndCache))
   }
 
   /** Transforms a String into an Validation[Throwable, InetAddress] */
-  private def getIpAddress(ip: String): Validation[Throwable, InetAddress] =
-    Validation.fromTryCatch(InetAddress.getByName(ip))
+  private def getIpAddress(ip: String): IO[Validation[Throwable, InetAddress]] =
+    IO { Validation.fromTryCatch(InetAddress.getByName(ip)) }
+
+  private def getCityResponse(
+    gs: DatabaseReader,
+    ipAddress: InetAddress
+  ): IO[Validation[Throwable, CityResponse]] =
+    IO { Validation.fromTryCatch(gs.city(ipAddress)) }
+
+  private def getFromMap[K, V](map: SynchronizedLruMap[K, V], k: K): IO[Option[V]] =
+    IO { map.get(k) }
```

Comment:
what's the reasoning behind wrapping `get` in IO?

Reply:
If `put` is called between calls to `get` it might return a different value so it's not referentially transparent. When I work on #42 maybe we could switch to an immutable map so get and set can be pure?